From fac62d7a4b338fccc5facb3dddfb14580a25e26e Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 16 Feb 2022 08:26:30 +0200
Subject: [PATCH] [SW-0] gaudi3: update dma api to remove size constraints

Currently the new dma API allows maximum DMA size of 128KB assuming
ppboot and preboot size wont cross 128KB. However in gaudi3, there
is possiblility that preboot size will cross 128KB. This patch
updates api to support any size dma provided the size and addresses
are aligned. Also api is modified to skip crc check based on function
argument.

Change-Id: Iabdffaa79820b8a12dfe68d34c398e0945286179
---

diff --git a/arch/gaudi3/arc/dev_dma.h b/arch/gaudi3/arc/dev_dma.h
index fdbb7a1..fb5dcc5 100644
--- a/arch/gaudi3/arc/dev_dma.h
+++ b/arch/gaudi3/arc/dev_dma.h
@@ -6,14 +6,13 @@
 #include <secureboot.h>
 
 #define MAX_DMA_SIZE		0xFFF0
-#define MAX_IMAGE_SIZE		(2 * (MAX_DMA_SIZE + CRC32_SIZE))
+#define MAX_DMA_SIZE_CRC	(MAX_DMA_SIZE + CRC32_SIZE)
 #define OFFSET_64KB		0x10000
 #define ADDR_SIZE_ALIGNMENT	16	/* 16 bytes */
 
 enum dma_error {
 	DMA_SUCCESS = 0,
 	DMA_INVALID_CRC_ERROR,
-	DMA_INVALID_IMG_SIZE,
 	DMA_INVALID_DEST_TYPE,
 	DMA_UNALIGNED_IMG_SIZE,
 	DMA_UNALIGNED_DEST_ADDR,
@@ -32,8 +31,10 @@
  * @flash_offset: SPI flash offset
  * @dest_addr: Full 64 bit destination address
  * @dest_type: Destination HBW: 0 SRAM: 1
+ * @skip_crc_check: whether to skip CRC check
  *
  * Return: enum dma_error
  */
 enum dma_error dma_copy_image_v2(uint32_t image_size, uint32_t flash_offset,
-				 uint64_t dest_addr, enum dma_dest_type dest_type);
+				 uint64_t dest_addr, enum dma_dest_type dest_type,
+				 bool skip_crc_check);
diff --git a/arch/gaudi3/arc/dma.c b/arch/gaudi3/arc/dma.c
index c6ea038..2f8844d 100644
--- a/arch/gaudi3/arc/dma.c
+++ b/arch/gaudi3/arc/dma.c
@@ -9,12 +9,12 @@
 #include <dev_dma.h>
 
 enum dma_error dma_copy_image_v2(uint32_t image_size, uint32_t flash_offset,
-				 uint64_t dest_addr, enum dma_dest_type dest_type)
+				 uint64_t dest_addr, enum dma_dest_type dest_type,
+				 bool skip_crc_check)
 {
-	uint32_t payload_size, copy_size;
-
-	if (image_size > MAX_IMAGE_SIZE)
-		return DMA_INVALID_IMG_SIZE;
+	uint32_t copy_size, dma_size;
+	uint32_t dma_chunks;
+	uint8_t i;
 
 	if (dest_type >= DMA_DEST_MAX)
 		return DMA_INVALID_DEST_TYPE;
@@ -22,33 +22,32 @@
 	if (dest_addr % ADDR_SIZE_ALIGNMENT)
 		return DMA_UNALIGNED_DEST_ADDR;
 
-	if (image_size <=  (MAX_DMA_SIZE + CRC32_SIZE)) {
-		if ((image_size - CRC32_SIZE) % ADDR_SIZE_ALIGNMENT)
-			return DMA_UNALIGNED_IMG_SIZE;
+	dma_chunks = image_size / MAX_DMA_SIZE_CRC;
+	if (image_size % MAX_DMA_SIZE_CRC)
+		dma_chunks++;
 
-		dma_copy_image(image_size, flash_offset, dest_addr, dest_type);
-		if (dma_wait_finish() == DMA_CRC_ERROR)
-			return DMA_INVALID_CRC_ERROR;
-		else
-			return DMA_SUCCESS;
-	}
-
-	payload_size = image_size - (2 * CRC32_SIZE);
-	if (payload_size % ADDR_SIZE_ALIGNMENT)
+	if ((image_size - (CRC32_SIZE * dma_chunks)) % ADDR_SIZE_ALIGNMENT)
 		return DMA_UNALIGNED_IMG_SIZE;
 
-	/* copy 1st chunk flash_off to dest_off */
-	copy_size = MAX_DMA_SIZE + CRC32_SIZE;
-	dma_copy_image(copy_size, flash_offset, dest_addr, dest_type);
-	if (dma_wait_finish() == DMA_CRC_ERROR)
-		return DMA_INVALID_CRC_ERROR;
+	copy_size = image_size;
+	/*
+	 * DMA example:
+	 * image_size = 200kb + 16 bytes CRC = 0x32010
+	 * dma_chunks: 4
+	 * 1st DMA: SRC: 0x0 DST: 0x0 (dma_size: 0xFFF4 left_size: 0x2201C)
+	 * 2nd DMA: SRC: 0x10000 DST: 0xFFF0 (dma_size: 0xFFF4 left_size: 0x12028)
+	 * 3nd DMA: SRC: 0x20000 DST: 0x1FFE0 (dma_size: 0xFFF4 left_size: 0x2034)
+	 * 4th DMA: SRC: 0x30000 DST: 0x2FFD0 (dma_size: 0x2034 left_size: 0x0)
+	 */
+	for (i = 0; i < dma_chunks; i++) {
+		dma_size = copy_size > MAX_DMA_SIZE_CRC ? MAX_DMA_SIZE_CRC : copy_size;
+		dma_copy_image(dma_size, (flash_offset + (OFFSET_64KB * i)),
+				(dest_addr + (MAX_DMA_SIZE * i)), dest_type);
+		if ((!skip_crc_check) && (dma_wait_finish() == DMA_CRC_ERROR))
+			return DMA_INVALID_CRC_ERROR;
 
-	/* copy 2nd chunk from flash_off + 0x10000 to dest_off + MAX_DMA_SIZE */
-	copy_size = payload_size - MAX_DMA_SIZE + CRC32_SIZE;
-	dma_copy_image(copy_size, (flash_offset + OFFSET_64KB),
-			(dest_addr + MAX_DMA_SIZE), dest_type);
-	if (dma_wait_finish() == DMA_CRC_ERROR)
-		return DMA_INVALID_CRC_ERROR;
+		copy_size -= dma_size;
+	}
 
 	return DMA_SUCCESS;
 }
